NDTensors/src/lib/GradedAxes/src/fusion.jl

```diff
@@ -83,24 +105,38 @@ end
 # Get the permutation for sorting, then group by common elements.
 # groupsortperm([2, 1, 2, 3]) == [[2], [1, 3], [4]]
 function blockmergesortperm(a::AbstractBlockedUnitRange)
-  # If it is dual, reverse the sorting so the sectors
-  # end up sorted in the same way whether or not the space
-  # is dual.
-  # TODO: Figure out how to deal with dual sectors.
-  # TODO: `rev=isdual(a)`  may not be correct for symmetries beyond `U(1)`.
-  ## return Block.(groupsortperm(nondual_sectors(a); rev=isdual(a)))
   return Block.(groupsortperm(blocklabels(a)))
 end
 
 # Used by `TensorAlgebra.splitdims` in `BlockSparseArraysGradedAxesExt`.
 invblockperm(a::Vector{<:Block{1}}) = Block.(invperm(Int.(a)))
 
-# Used by `TensorAlgebra.fusedims` in `BlockSparseArraysGradedAxesExt`.
-function blockmergesortperm(a::GradedUnitRange)
-  # If it is dual, reverse the sorting so the sectors
-  # end up sorted in the same way whether or not the space
-  # is dual.
-  # TODO: Figure out how to deal with dual sectors.
-  # TODO: `rev=isdual(a)`  may not be correct for symmetries beyond `U(1)`.
-  return Block.(groupsortperm(blocklabels(a)))
+function blockmergesortperm(a::UnitRangeDual)
+  return Block.(groupsortperm(blocklabels(nondual(a))))
+end
+
+function blockmergesort(g::AbstractGradedUnitRange)
+  glabels = blocklabels(g)
+  gblocklengths = blocklengths(g)
+  new_blocklengths = map(
+    la -> labelled(sum(gblocklengths[findall(==(la), glabels)]; init=0), la),
+    sort(unique(glabels)),
+  )
+  return GradedAxes.gradedrange(new_blocklengths)
+end
+
+blockmergesort(g::UnitRangeDual) = dual(blockmergesort(flip(g)))
+blockmergesort(g::OneToOne) = g
+
+# fusion_product produces a sorted, non-dual GradedUnitRange
+function fusion_product(g1, g2)
+  return blockmergesort(tensor_product(g1, g2))
+end
+
+fusion_product(g::AbstractUnitRange) = blockmergesort(g)
+fusion_product(g::UnitRangeDual) = fusion_product(flip(g))
+
+# recursive fusion_product. Simpler than reduce + fix type stability issues with reduce
+function fusion_product(g1, g2, g3...)
+  return fusion_product(fusion_product(g1, g2), g3...)
```

Comment:
I was going to ask why this wasn't defined as:
```julia
  return foldl(fusion_product, (g1, g2, g3...))
```
Note it is better to use `foldl` rather than `reduce` in cases like this where you want to specify the order of operations.

Reply:
I just checked again, with `reduce` or `foldl` julia is sometimes unable to predict the output type when non-abelian irreps are fused. I would rather keep the current version.

Reply:
That's too bad, sounds reasonable to keep it as is in that case.